File: api-chi/cmd/services/movie.go
package services

import (
	"api-chi/cmd/config"
	"api-chi/cmd/models"

	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

type MovieService struct {
	Conn *pgxpool.Pool
}

func (s *MovieService) Open() error {
	config.LoadConfig()
	conn, err := pgxpool.New(config.CTX, config.POSTGRES_URL)
	if err != nil {
		return err
	}

	// Assign connection
	s.Conn = conn
	return nil
}

func (s *MovieService) Close() {
	s.Conn.Close()
}

func (s *MovieService) Create(input *models.Movie) (string, error) {
	// Execute SQL
	sql := "SELECT * FROM create_movie(@name, @initial_date, @thumbnail_image, @trailer_video_url, @description);"
	value := ""
	args := pgx.NamedArgs{
		"name":              input.Name,
		"initial_date":      input.InitialDate,
		"thumbnail_image":   input.ThumbnailImage,
		"trailer_video_url": input.TrailerVideoUrl,
		"description":       input.Description,
	}
	if input.Name == "" {
		args["name"] = nil
	}
	if input.InitialDate == "" {
		args["initial_date"] = nil
	}
	if input.ThumbnailImage == "" {
		args["thumbnail_image"] = nil
	}
	if input.TrailerVideoUrl == "" {
		args["trailer_video_url"] = nil
	}
	if input.Description == "" {
		args["description"] = nil
	}

	err := s.Conn.QueryRow(config.CTX, sql, args).Scan(&value)
	if err != nil {
		return value, err
	}

	// If success return nil
	return value, nil
}

func (s *MovieService) Update(input *models.Movie) (models.Movie, error) {
	// Execute SQL
	sql := "SELECT * FROM update_movie(@id, @name, @initial_date, @thumbnail_image, @trailer_video_url, @description);"
	args := pgx.NamedArgs{
		"id":                input.Id,
		"name":              input.Name,
		"initial_date":      input.InitialDate,
		"thumbnail_image":   input.ThumbnailImage,
		"trailer_video_url": input.TrailerVideoUrl,
		"description":       input.Description,
	}
	if input.Name == "" {
		args["name"] = nil
	}
	if input.InitialDate == "" {
		args["initial_date"] = nil
	}
	if input.ThumbnailImage == "" {
		args["thumbnail_image"] = nil
	}
	if input.TrailerVideoUrl == "" {
		args["trailer_video_url"] = nil
	}
	if input.Description == "" {
		args["description"] = nil
	}

	value := models.Movie{}
	valueTime := time.Time{}
	err := s.Conn.QueryRow(config.CTX, sql, args).Scan(
		&value.Id,
		&value.Name,
		&valueTime,
		&value.ThumbnailImage,
		&value.TrailerVideoUrl,
		&value.Description,
	)
	if err != nil {
		return value, err
	}
	value.InitialDate = valueTime.Format("2006-01-02")

	// If success return nil
	return value, nil
}

func (s *MovieService) Remove(id *string) (string, error) {
	// Execute SQL
	sql := "SELECT * FROM remove_movie(@id);"
	args := pgx.NamedArgs{
		"id": *id,
	}
	value := ""
	err := s.Conn.QueryRow(config.CTX, sql, args).Scan(&value)
	if err != nil {
		return value, err
	}

	// If success return nil
	return value, nil
}

func (s *MovieService) Count() (int, error) {
	// Execute SQL
	sql := "SELECT * FROM count_movie();"
	value := 0
	err := s.Conn.QueryRow(config.CTX, sql).Scan(&value)
	if err != nil {
		return value, err
	}

	// If success return nil
	return value, nil
}

func (s *MovieService) GetAll(search *string, limit *int, page *int) ([]models.Movie, error) {
	// Set default range for limit
	if *limit < 10 {
		*limit = 10
	} else if *limit > 50 {
		*limit = 50
	}

	// Set default range for page
	if *page < 1 {
		*page = 0
	} else {
		*page = *page - 1
	}

	// Execute SQL
	sql := "SELECT * FROM get_all_movie(@search, @limit, @page);"
	args := pgx.NamedArgs{
		"search": *search,
		"limit":  *limit,
		"page":   *page,
	}
	value := []models.Movie{}
	rows, err := s.Conn.Query(config.CTX, sql, args)
	if err != nil {
		return value, err
	}
	for rows.Next() {
		item := models.Movie{}

		if err := rows.Scan(&item.Id, &item.Name); err != nil {
			return nil, err
		}

		value = append(value, item)
	}

	// If success return nil
	return value, nil
}

File: api-fastify/src/api/services/movie.ts
import { EnumMessage, ReturnMessage } from "@/types/message"
import pool from "@/libs/config/database_postgres"
import {  Movie } from "@/types/movie"

export default class MovieService {
  public async Create(input: Movie): Promise<ReturnMessage> {
    return pool
      .withClient(async (client) => {
        const result = await client.query(
          "SELECT * FROM create_movie($1, $2, $3, $4, $5)",
          [
            input.name,
            input.initialDate,
            input.thumbnailImage,
            input.trailerVideoUrl,
            input.description,
          ],
        )

        if (!Array.isArray(result.rows) || result.rows.length === 0)
          throw new Error("Created failed, no movie returned.")

        const row = result.rows[0]
        const movie: Movie = {
          id: row.id,
          name: row.name,
          initialDate: new Date(row.initial_date),
          thumbnailImage: row.thumbnail_image,
          trailerVideoUrl: row.trailer_video_url,
          description: row.description,
        }

        return { success: true, message: EnumMessage.CREATE_SUCCESS, data: movie }
      })
      .catch((error) => this.handleError(error, EnumMessage.CREATE_FAILED))
  }

  public async Update(input: Movie): Promise<ReturnMessage> {
    return pool
      .withClient(async (client) => {
        const result = await client.query(
          "SELECT * FROM update_movie($1, $2, $3, $4, $5, $6)",
          [
            input.id,
            input.name,
            input.initialDate,
            input.thumbnailImage,
            input.trailerVideoUrl,
            input.description,
          ],
        )

        if (!Array.isArray(result.rows) || result.rows.length === 0)
          throw new Error("Update failed, no movie returned.")

        const row = result.rows[0]
        const movie: Movie = {
          id: row.id,
          name: row.name,
          initialDate: new Date(row.initial_date),
          thumbnailImage: row.thumbnail_image,
          trailerVideoUrl: row.trailer_video_url,
          description: row.description,
        }

        return {
          success: true,
          message: EnumMessage.UPDATE_SUCCESS,
          data: movie,
        }
      })
      .catch((error) => this.handleError(error, EnumMessage.UPDATE_FAILED))
  }

  public async Remove(id: string): Promise<ReturnMessage> {
    return pool
      .withClient(async (client) => {
        const result = await client.query(
          "SELECT * FROM remove_movie($1)",
          [id],
        )
        const removeId: string = result.rows[0].remove_movie
        return {
          success: true,
          message: EnumMessage.REMOVE_SUCCESS,
          data: removeId,
        }
      })
      .catch((error) => this.handleError(error, EnumMessage.REMOVE_FAILED))
  }

  public async Count(): Promise<ReturnMessage> {
    return pool
      .withClient(async (client) => {
        const result = await client.query(
          "SELECT * FROM count_movie()",
        )
        const data: number = parseInt(result.rows[0].count_movie)
        return { success: true, message: EnumMessage.GET_SUCCESS, data: data }
      })
      .catch((error) => this.handleError(error, EnumMessage.GET_FAILED))
  }

  public async GetAll(
    search: string,
    limit: number,
    page: number,
  ): Promise<ReturnMessage> {
    return pool
      .withClient(async (client) => {
        const result = await client.query(
          "SELECT * FROM get_all_movie($1, $2, $3)",
          [search, limit, page - 1],
        )

        let data: Movie[] = Array.isArray(result.rows)
          ? result.rows.map((row) => {
              return {
                id: row.id,
                name: row.name,
                initialDate: new Date(row.initial_date),
                thumbnailImage: row.thumbnail_image,
                trailerVideoUrl: row.trailer_video_url,
                description: row.description,
              } as Movie
            })
          : []

        return { success: true, message: EnumMessage.GET_SUCCESS, data: data }
      })
      .catch((error) => this.handleError(error, EnumMessage.GET_FAILED))
  }

  private handleError(error: any, message: EnumMessage): ReturnMessage {
    console.error(error)
    return {
      success: false,
      message,
      data: undefined,
      error_detail: error.message,
    }
  }
}
